### DISCUSSION TITLE
Desktop support

### DISCUSSION THREAD

**Discussion by Sesa1988** [auth0/auth0-flutter#802] [General]
### Checklist

- [X] I have looked into the [README](https://github.com/auth0/auth0-flutter/tree/main/auth0_flutter#readme) and have not found a suitable solution or answer.
- [X] I have looked into the [API documentation](https://pub.dev/documentation/auth0_flutter/latest/) and have not found a suitable solution or answer.
- [X] I have searched the [issues](https://github.com/auth0/auth0-flutter/issues) and have not found a suitable solution or answer.
- [X] I have searched the [Auth0 Community](https://community.auth0.com/c/sdks/5) forums and have not found a suitable solution or answer.
- [X] I agree to the terms within the [Auth0 Code of Conduct](https://github.com/auth0/open-source-template/blob/master/CODE-OF-CONDUCT.md).

### Describe the problem you'd like to have solved

Desktop support especially with social providers would be awesome, there is currently no package that supports this.

### Describe the ideal solution

Everything should work like now but there is probably more work to be done because windows does not support url redirects.
Other packages start a http server and listen for callbacks but they are all specific to a product.

**Comment by Widcket** [comment]
Hi @Sesa1988, thanks for raising this.

Is this feature request for Windows support specifically?

**Comment by Sesa1988** [comment]
> Hi @Sesa1988, thanks for raising this.
> 
> Is this feature request for Windows support specifically?

This was just my preference but web support would be nice as well.

Im looking for a auth package that can be used on all Flutter platforms that support azure ad b2c and similar services.

**Comment by Widcket** [comment]
We already have https://github.com/auth0/auth0-flutter/issues/169 for tracking web support. As mentioned there:

> Since Flutter for web supports interop with JS, you could use our SDK for SPAs: https://github.com/auth0/auth0-spa-js/.

Would that work for you?

**Comment by Sesa1988** [comment]
> We already have https://github.com/auth0/auth0-flutter/issues/169 for tracking web support. As mentioned there:
> 
> > Since Flutter for web supports interop with JS, you could use our SDK for SPAs: https://github.com/auth0/auth0-spa-js/.
> 
> Would that work for you?

Im not sure how well this plays with Flutter web but Im looking more for desktop support.

**Comment by Widcket** [comment]
Thanks. We have no plans in the near term but I'll leave this issue open for future planning and reference.

**Comment by sir-dunxalot** [comment]
+1 to providing desktop support or making is clearer in the readme, tutorials, and quickstart that desktop is not supported.

Alternatively, it would be helpful to have a clearer error message. Trying to run auth0-flutter in a MacOS app results in the following error, which is misleading:

```
MissingPluginException(No implementation found for method webAuth#login on channel auth0.com/auth0_flutter/web_auth)
```

Thanks!

**Comment by stevehobbsdev** [comment]
@sir-dunxalot I think that's an error surfaced by the Flutter runtime rather than our plugin.

We're always happy to look at ways to make our support clearer. Currently the [package on pub.dev](https://pub.dev/packages/auth0_flutter) is accurate in that it shows iOS and Android as supported platforms:

<img width="531" alt="image" src="https://user-images.githubusercontent.com/766403/221166815-68d54a70-32ad-4533-b5f1-e0cb581bcc2a.png">

And we mention supported in the readme for this repo:

<img width="619" alt="image" src="https://user-images.githubusercontent.com/766403/221167044-33b4e5f2-f22a-4267-b23e-f361c82cb4c4.png">

**Comment by keepersmith** [comment]
+1 to providing desktop support.
One of the main reasons to use flutter in the first place is to chase the dream of one codebase for all (relevant) platforms.
Please add desktop (and web) support.
Thanks!

**Comment by steve-winter** [comment]
As I see it there are a few options:

**Loopback:** - potential MiTM risks
  1. Login triggers opening browser
  2. Perform normal login in browser
  3. Run server on loopback address
  4. Browser redirects to a loopback address (e.g. 127.0.0.1:1234)
  5. Close browser
  6. Loopback address broadly follows; https://github.com/auth0/auth0-flutter/blob/main/auth0_flutter/lib/src/web/auth0_flutter_plugin_real.dart 
  7. Close loopback server and listen for login in initState. 

**Embedded Webview**
1. Open login in embedded webview
3. Listen for relevant redirect and extract values

~~**MacOS only**~~
~~1. Package redirect (e.g. https://auth0.com/docs/quickstart/native/ios-swift/interactive)~~

Keen for any feedback re recommended/agreeable direction, and I will begin to put a PR together.

**Comment by Xylez01** [comment]
Any updates on this topic? We are currently using https://pub.dev/packages/flutter_appauth (as it was used in previous Auth0 blogs on how to integrate with Auth0 and flutter).

Since then macOS support has been added which we are using quite intensely. Do you have any idea on the timeline for macOS support for this lib? I think ideally we'd like to move over to this lib but no macOS support is blocking us currently.

**Comment by raphibolliger** [comment]
For me, it would be also really nice to have Auth0 support on desktop platforms. Any chance that this will be implemented in the near future? Especially macOS should be almost the same as iOS.

**Comment by Anish-Chanda** [comment]
+1

**Comment by Xylez01** [comment]
@Widcket are there any plans currently to support this?

**Comment by Widcket** [comment]
Hi @JordyLangen, we are currently working on adding support for macOS, though I can't share a specific ETA yet.

**Comment by Xylez01** [comment]
@Widcket thats good to hear! Do you have a time window? This year? Q1 next year?

**Comment by Widcket** [comment]
Though this is no ETA, a rough time window would be around end of year.

**Comment by Widcket** [comment]
Hey everyone, macOS support is out now on [v1.5.0](https://github.com/auth0/auth0-flutter/releases/tag/af-v1.5.0).

**Comment by jaipal-sodhi** [comment]
Please provide support for windows as well

**Comment by Gaurav-CareMonitor** [comment]
@Widcket  Any updates on Windows support? Is this going to be added in the near future?

**Comment by avinash-bandgitech** [comment]
Any updates on Windows support?

**Comment by TristanGrobler** [comment]
I would also love windows support.

**Comment by emal-avala** [comment]
@Widcket is there a timeline for Windows support? Thanks in advance!

**Comment by Widcket** [comment]
Hi @emal-avala, there are no current plans for Windows support.

**Comment by esDotDev** [comment]
Why are you not intending on supporting Windows? Is it a technical reason?

**Comment by cmarques2** [comment]
Am looking to create a windows app that uses Auth0. Any progress/update on this would be amazing.

**Comment by esDotDev** [comment]
Can we get some sort of clarification on this? You're missing out on over 1 billion users by not supporting windows. 

Additionally when building mobile apps testing responsiveness with a resizable desktop window is a massive productivity booster, which is also being blocked.

@Widcket any thoughts?

**Comment by Widcket** [comment]
cc @pmathew92 you might want to bring this up to the PM

**Comment by pmathew92** [comment]
Hi folks, will be discussing this internally with our team to check when we can include this in our roadmap. Thank you for your patience

**Comment by esDotDev** [comment]
Thanks for the update!

**Comment by esDotDev** [comment]
> Hi folks, will be discussing this internally with our team to check when we can include this in our roadmap. Thank you for your patience

Any clarification on this? We're trying to decide whether we need to devise our own workaround, or if we can potentially wait for windows support in this lib.

**Comment by pmathew92** [comment]
Hi everyone, apologies on the delayed response. We plan to pick this up in the  third quarter.

**Comment by esDotDev** [comment]
Thanks for the clarity, much appreciated.

**Comment by hkroger** [comment]
@pmathew92 any updates on the timeline?

**Comment by MrTadeu** [comment]
Hi everyone, I saw Windows desktop support was mentioned for Q3. I also lost some time due to the “Desktop” wording (please change to only macOS). Are there any updates on the Windows timeline? Thanks!

"Mobile/Desktop" -> "Mobile/MacOS"

<img width="480" height="270" alt="Image" src="https://github.com/user-attachments/assets/e7298784-3a41-45a5-9398-1947fca5a0a1" />

**Comment by NandanPrabhu** [comment]
@hkroger @MrTadeu  macOS support is already available as part of the plugin. Windows support will take some more time. We are actively working on it in current sprint. We are aiming to release Windows support by end of Q3

**Comment by Vild** [comment]
I've published a package that extends the platform interface to add Linux and Windows support. Which is required for project that I'm working on.

Should be pretty easy to add to project, as all you need to do is call initDesktopAuth0Flutter on windows and linux.
After that auth0_flutter works as expected :tm:.

In-case anyone needs it: 
https://pub.dev/packages/desktop_auth0_flutter

**Comment by Dispersia** [comment]
Sorry to badger, but is there any update on this? I will look at what Vild did, but built in would be nice

**Comment by NandanPrabhu** [comment]
Hi @Dispersia, the feature request has been worked on in the current quarter. We will update you with a confirmed release date soon.

**Comment by eivihnd** [comment]
Really looking forward to enabling Auth0 login in our Windows Flutter app. How is the work progressing?
Will it be part of the new v2?

**Comment by NandanPrabhu** [comment]
Hi @eivihnd , work is in progress. Code review and security review will be completed this week. We will be releasing first beta next week. v2.0 wont have flutter desktop support but v2.1-beta would

**Comment by NandanPrabhu** [comment]
Hi Everyone, We have some exciting news. We have released auth0-flutter v2.1.0-beta.1 adding windows desktop support [v2.1.0-beta.1](https://github.com/auth0/auth0-flutter/releases/tag/af-v2.1.0-beta.1). It currently supports Web authentication and we will be adding more features to it and also stabilization fixes as we close the beta release and go GA. Please try it out and share your feedback in our discussion forum or [Auth0 support channel](https://support.auth0.com/center/s/)

**Comment by NandanPrabhu** [comment]
Flutter desktop is GA now with [v2.1.0](https://github.com/auth0/auth0-flutter/releases/tag/af-v2.1.0) release 🚀. Please try it out and let us know your feedback!

**Comment by eivihnd** [comment]
Thank you so much for adding Windows support! Implementing it now, and having the following issue:

The `hasValidCredentials` in `CredentialsManager` throws 
`MissingPluginException(No implementation found for method credentialsManager#hasValidCredentials on channel auth0.com/auth0_flutter/credentials_manager)`

I do realize that the [docs](https://github.com/auth0/auth0-flutter/blob/main/auth0_flutter/EXAMPLES.md#1-install-vcpkg-and-native-dependencies) says I have to handle storing the credentials myself. What are the reason behind this?

The `hasValidCredentials` available on MacOS and mobile is certainly a convenient function!

**Reply by NandanPrabhu** [reply]
Hi @eivihnd, since we wanted to first release web authentication and gather feedback, the first release only has web authentication. Credentials management is something which we are working on and you can expect a release on the same in a sprint's time.

**Comment by Dispersia** [comment]
ya, I ran into it as well, I added that and linux support to my fork. If I could make a request, cpprestsdk had already been deprecated, and is incredibly hard to build. Is it possible to use something more like httplib.h, which is just a drop-in header file that doesn't need external dependencies to build? Cpprestsdk and Boost are extremely expensive for CI servers. 

 Definitely appreciate the effort put into it, and excited to see what it becomes

**Reply by NandanPrabhu** [reply]
yes @Dispersia. We released first version with cpp rest sdk and boost but it doesn't stop here. We are exploring other options to optimise it further and definitely httplib is one of them. We will be taking networking layer optimisation as well as adding credential management together as part of new release. I cannot commit an exact ETA but we are prioritising implementation next week. You can expect it to be out by end of June

**Reply by Dispersia** [reply]
That would be great, not having to maintain a fork and using just the official implementation would be fantastic. Thank you!
